### PR TITLE
producer: add support for deduplication

### PR DESCRIPTION
**Pull request by gvolpe**

### CODE REVIEW

**Comment by gvolpe** [comment]
No idea why the new test keeps on failing on the CI. It passes 10/10 times locally 🤷🏽 Could it be GH actions doesn't reflect the changes in the script and docker-compose configuration from the PR? Merging to find out...

File: core/src/main/scala/dev/profunktor/pulsar/Producer.scala [hunk 1]
```diff
@@ -126,30 +138,75 @@ object Producer {
...
+                  .withMessageKey(key)
+              ).sendAsync()
+            }
+
+      settings.deduplication match {
+        case Deduplication.Enabled =>
+          prevMsgs.modify { prev =>
+            val nextId = SeqIdMaker[E].next(p.getLastSequenceId(), prev, msg)
+            Some(msg) -> cont(_.sequenceId(nextId))
+          }.flatten
```
Comment (Daenyth) [inline core/src/main/scala/dev/profunktor/pulsar/Producer.scala:204]: I'm suspicious about this - I wonder if the test failure is a race condition here

Can you make it stop flaking if you slap a semaphore(1) on this section

Comment (gvolpe) [reply core/src/main/scala/dev/profunktor/pulsar/Producer.scala:204]: Good eye! I can actually move the `cont` effect outside the `modify` block and see if that helps...

Comment (gvolpe) [reply core/src/main/scala/dev/profunktor/pulsar/Producer.scala:204]: Changed it to this in #20, no luck

```scala
prevMsgs
  .modify { prev =>
    val nextId = seqIdMaker
      .asInstanceOf[SeqIdMaker[E]]
      .next(p.getLastSequenceId(), prev, msg)
    Some(msg) -> nextId
  }
  .flatMap(nextId => cont(_.sequenceId(nextId)))
```

I'm going to add a log for that `nextId` to see if it's the right one

Comment (gvolpe) [reply core/src/main/scala/dev/profunktor/pulsar/Producer.scala:204]: Doesn't seem to be this piece of code :( https://github.com/profunktor/neutron/runs/5725075664?check_suite_focus=true

Comment (Daenyth) [reply core/src/main/scala/dev/profunktor/pulsar/Producer.scala:204]: I don't think changing `flatten` to `flatMap` changes the actual behavior *at all* though - it just changes how the method is structured.

In both cases, we have `modify state >> async interruption point >> execute cont`. I know it's probably not reasonable to have the final production code using a lock here, but I'd still check with a Semaphore

Comment (gvolpe) [reply core/src/main/scala/dev/profunktor/pulsar/Producer.scala:204]: Right, I thought the `cont` effect within the `modify` could cause some contention, but you're right the behavior remains the same.

So are you suggesting the issue could be an interruption before `cont(_.sequenceId(nextId))` is executed? 🤔

Comment (gvolpe) [reply core/src/main/scala/dev/profunktor/pulsar/Producer.scala:204]: Added a semaphore in [this commit](https://github.com/profunktor/neutron/pull/20/commits/793146e6fdeec069714d61b6fdfe5fc34aa1d9a9), still same results 😭

Comment (Daenyth) [reply core/src/main/scala/dev/profunktor/pulsar/Producer.scala:204]: Why are concurrency problems never solved by the easy thing? :sigh:

Comment (gvolpe) [reply core/src/main/scala/dev/profunktor/pulsar/Producer.scala:204]: I doubt this is a concurrency issue, though. Looks more like a Pulsar discrepancy on CI... But I'm using the same Docker image :/

**Pull request merged**